lib/mix/tasks/benchmark.ex

```diff
@@ -0,0 +1,199 @@
+defmodule Mix.Tasks.Benchmark.Helper do
+  @moduledoc false
+  defmacro compile_file(file) do
+    case compare_versions(System.version(), "1.7.0") do
+      :gt ->
+        quote do
+          Code.compile_file(unquote(file))
+        end
+
+      _ ->
+        quote do
+          Code.load_file(unquote(file))
+        end
+    end
+  end
+
+  defp compare_versions(v1, v2) do
+    Version.compare(Version.parse!(v1), Version.parse!(v2))
+  end
+end
+
+defmodule Mix.Tasks.Benchmark do
+  use Mix.Task
+  require Logger
+  alias Mix.Tasks.Benchmark.Helper
+  require Mix.Tasks.Benchmark.Helper
+
+  @moduledoc """
+  Runs project benchmarks
+  """
+  @dialyzer [no_match: [find_benchmark_files: 2]]
+
+  @switches []
+
+  @shortdoc "Runs project benchmarks"
+  @recursive true
+  @preferred_cli_env :test
+  @impl Mix.Task
+  def run(args) do
+    {opts, files} = OptionParser.parse!(args, strict: @switches)
+
+    compile_project(args, opts)
+
+    files
+    |> compile_benchmarks(opts)
+    |> run_benchmarks(opts)
+  end
+
+  defp compile_project(args, _opts) do
+    :ok = Logger.debug("Compiling project")
+    Mix.Task.run("loadpaths", args)
+    Mix.Project.compile(args)
+  end
+
+  defp compile_benchmarks(files, opts) do
+    files
+    |> find_benchmark_files(opts)
+    |> compile_benchmark_files(opts)
+  end
+
+  defp find_benchmark_files(:stop, _opts), do: :stop
+
+  defp find_benchmark_files(files, _opts) do
+    :ok = Logger.debug("Locating benchmarks")
+
+    files =
+      case files do
+        [] ->
+          [default_benchmark_path()]
+
+        o ->
+          o
+      end
+
+    files =
+      Enum.flat_map(files, fn file ->
+        if File.dir?(file) do
+          [file, "*.exs"]
+          |> Path.join()
+          |> Path.wildcard()
+        else
+          [file]
```

Comment:
just `file` should be enough here shouldn't it?

Reply:
The function must return list of files found for the further use in iterator functions.

Reply:
Sure? It's being flat_maped, so the wrapped list should be removed anyhow.

Reply:
Uh, not sure if it was a question, but here's a piece of iex session:
```
iex(1)> [:a, :b, :c] |> Enum.flat_map(&[&1])
[:a, :b, :c]
iex(2)> [:a, :b, :c] |> Enum.flat_map(& &1)
** (Protocol.UndefinedError) protocol Enumerable not implemented for :a of type Atom
```

Reply:
I think I meant more because we return `[file]` here and alter `flat_map` why can't we just return `file` and then `map` instead of `flat_map`?